plugin/pkg/scheduler/schedulercache/interface.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulercache
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+// Cache collects pods' information and provides node-level aggregated information.
+// It's intended for generic scheduler to do efficient lookup.
+// Cache's operations are pod centric. It incrementally updates itself based on pod event.
+// Pod events are sent via network. We don't have guaranteed delivery of all events.
+// Thus, we organized the state machine flow of a pod's events as followed.
+//
+// State Machine of a pod's events in scheduler's cache:
+//
+//                                                +-------+
+//                                                |       |
+//                                                |       | Update
+//           Assume                Add            +       |
+// Initial +--------> Assumed +------------+---> Added <--+
+//                       +                 |       +
+//                       |                 |       |
+//                       |             Add |       | Remove
+//                       |                 |       |
+//                       |                 +       |
+//                       +-------------> Expired   +----> Deleted
+//                          expire
+//
+// Note that an assumed pod would be expired. Because if we haven't received Add event
+// notifying us that it's scheduled, there might be some problems and we shouldn't assume
+// the pod scheduled anymore.
+//
+// Note that "Initial", "Expired", and "Deleted" pods do not actually exist in cache.
+// Based on existing use cases, we are making following assumptions:
+// - No same pod would be assumed twice
+// - If a pod wasn't added before, it wouldn't be removed or updated.
+// - Both "Expired" and "Deleted" are valid end states. An expired pod could never
+//   be added if it missed all events due to network disconnection.
```

Comment:
I added a bunch of comments in the commit:
https://github.com/hongchaodeng/kubernetes/commit/2e7f039a4c9a5226d57b1be47c46c92dc5fe7a99